Messaging - DAO changes

kaysoky opened the pull request:
Adds two join tables for messages to threads and threads to objects.
Each message can belong to one thread and each thread can belong to one object.

kaysoky committed:
Add uniqueness constraint to message DTO (simplifies other logic)

kaysoky committed:
In the case where a message thread is tied to an entity (i.e. a comment thread), there shouldn't be any user-ID filtering.  The filtering should be on the entire thread, done on the manager level (i.e. does the user have READ permissions on the entity?)

kaysoky committed:
Add link between MessageThreads and Nodes

kaysoky committed:
Migration tests...

kaysoky committed:
Need a conversion between threads and nodes and vice versa.
+ Manager interface

kaysoky committed:
Threads should not be linked to nodes, but to objects
This means threads will have to be broken off from messages into a join table

kaysoky committed:
+ Fix up a few resulting errors
+ Revise DAO interface

kaysoky committed:
Fix up remainder of the tests

kaysoky committed:
Message threads can be a secondary type of messages since their unique keys are the same.
However, message thread objects can't be secondary to message threads, since the unique keys don't match.

kaysoky committed:
Need to save the reply-to information for asynchronous sending of messages

kaysoky committed:
Add null tolerance to utility

brucehoff commented on lib/jdomodels/src/main/resources/schema/Message-ddl.sql:
```diff
@@ -1,14 +1,14 @@
 CREATE TABLE `MESSAGE` (
   `ID` bigint(20) NOT NULL,
-  `THREAD_ID` bigint(20) NOT NULL,
```
consider removing thread_id from Message and just having reply_to.  We'd then have a separate "reporting" table with two columns, (1) message_id (FK to this table's ID) and (2) THREAD, also a FK to this table's ID).  The ID of the first message in a thread is the ID of the thread.

In the reporting table, the Thread column is a foreign key to the message column.

kaysoky committed:
Looks like foreign keys tied to non-unique keys is an InnoDB extension to MySQL.  When *any* row of the foreign key is deleted, the change will cascade delete every dependent row even if the foreign key constraint is still met (due to duplicate key).

We could have a single-column table just to hold threadIDs and thereby prevent this weirdness.

kaysoky committed:
Convert Message into MessageContent

kaysoky committed:
Convert remainder of Message object to a MessageToUser object

kaysoky committed:
Add the Message recipient object

kaysoky committed:
Add Message status and message in-reply-to
